File: core/src/nl/t64/game/rpg/screens/world/conversation/ConversationSubject.java
package nl.t64.game.rpg.screens.world.conversation;

import nl.t64.game.rpg.components.loot.Loot;

import java.util.ArrayList;
import java.util.List;


public class ConversationSubject {

    private final List<ConversationObserver> observers = new ArrayList<>();

    public void addObserver(ConversationObserver observer) {
        observers.add(observer);
    }

    public void removeObserver(ConversationObserver observer) {
        observers.remove(observer);
    }

    public void removeAllObservers() {
        observers.clear();
    }

    public void notifyExitConversation() {
        observers.forEach(ConversationObserver::onNotifyExitConversation);
    }

    public void notifyShowMessageTooltip(String message) {
        observers.forEach(observer -> observer.onNotifyShowMessageTooltip(message));
    }

    public void notifyShowLevelUpDialog(String message) {
        observers.forEach(observer -> observer.onNotifyShowLevelUpDialog(message));
    }

    public void notifyLoadShop() {
        observers.forEach(ConversationObserver::onNotifyLoadShop);
    }

    public void notifyShowRewardDialog(Loot reward, String levelUpMessage) {
        observers.forEach(observer -> observer.onNotifyShowRewardDialog(reward, levelUpMessage));
    }

    public void notifyShowReceiveDialog(Loot receive) {
        observers.forEach(observer -> observer.onNotifyShowReceiveDialog(receive));
    }

    public void notifyHeroJoined() {
        observers.forEach(ConversationObserver::onNotifyHeroJoined);
    }

    public void notifyHeroDismiss() {
        observers.forEach(ConversationObserver::onNotifyHeroDismiss);
    }

}

File: core/src/nl/t64/game/rpg/screens/world/conversation/ConversationSubject.kt
package nl.t64.game.rpg.screens.world.conversation

import nl.t64.game.rpg.components.loot.Loot


class ConversationSubject {

    private val observers: MutableList<ConversationObserver> = ArrayList()

    fun addObserver(observer: ConversationObserver) {
        observers.add(observer)
    }

    fun removeObserver(observer: ConversationObserver) {
        observers.remove(observer)
    }

    fun removeAllObservers() {
        observers.clear()
    }

    fun notifyExitConversation() {
        observers.forEach { it.onNotifyExitConversation() }
    }

    fun notifyShowMessageTooltip(message: String) {
        observers.forEach { it.onNotifyShowMessageTooltip(message) }
    }

    fun notifyShowLevelUpDialog(message: String) {
        observers.forEach { it.onNotifyShowLevelUpDialog(message) }
    }

    fun notifyLoadShop() {
        observers.forEach { it.onNotifyLoadShop() }
    }

    fun notifyShowRewardDialog(reward: Loot, levelUpMessage: String?) {
        observers.forEach { it.onNotifyShowRewardDialog(reward, levelUpMessage) }
    }

    fun notifyShowReceiveDialog(receive: Loot) {
        observers.forEach { it.onNotifyShowReceiveDialog(receive) }
    }

    fun notifyHeroJoined() {
        observers.forEach { it.onNotifyHeroJoined() }
    }

    fun notifyHeroDismiss() {
        observers.forEach { it.onNotifyHeroDismiss() }
    }

}
